types/share_splitting.go

```diff
@@ -2,11 +2,62 @@ package types
 
 import (
 	"bytes"
+	"fmt"
+	"sort"
 
 	"github.com/celestiaorg/nmt/namespace"
 	"github.com/tendermint/tendermint/pkg/consts"
 )
 
+// MessageShareWriter lazily merges messages into shares that will eventually be
+// included in a data square. It also has methods to help progressively count
+// how many shares the messages written take up.
+type MessageShareWriter struct {
+	shares [][]NamespacedShare
+	count  int
+}
+
+func NewMessageShareWriter() *MessageShareWriter {
+	return &MessageShareWriter{}
+}
+
+// Write adds the delimited data to the underlying contiguous shares
```

Comment:
nit: Consistently use punctuation everywhere. Or at least for public facing doc comments.

```suggestion
// Write adds the delimited data to the underlying contiguous shares.
```